src/crypto/common_functions.cpp

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <libp2p/crypto/common_functions.hpp>
+
+#include <gsl/gsl_util>
+#include <libp2p/crypto/error.hpp>
+
+namespace libp2p::crypto {
+
+  outcome::result<std::shared_ptr<EC_KEY>> EcKeyFromPrivateKeyBytes(
+      int nid, gsl::span<const uint8_t> private_key) {
+    auto FAILED = KeyGeneratorError::INTERNAL_ERROR;
+
+    // try initialize requested curve group
+    EC_GROUP *curve = EC_GROUP_new_by_curve_name(nid);
+    if (nullptr == curve) {
+      return CryptoProviderError::UNKNOWN_KEY_TYPE;
+    }
+    auto free_curve = gsl::finally([curve] { EC_GROUP_free(curve); });
+
+    // allocate the key
+    std::shared_ptr<EC_KEY> key{EC_KEY_new(), EC_KEY_free};
+    if (nullptr == key.get()) {
+      return FAILED;
+    }
+
+    // associate the curve with the key
+    if (1 != EC_KEY_set_group(key.get(), curve)) {
+      return FAILED;
+    }
+
+    // turn private key bytes into big number
+    BIGNUM *private_bignum{
+        BN_bin2bn(private_key.data(), private_key.size(), nullptr)};
+    if (nullptr == private_bignum) {
+      return FAILED;
+    }
+    auto free_private_bignum =
+        gsl::finally([private_bignum] { BN_free(private_bignum); });
+
+    // set private key to the resulting key structure
+    if (1 != EC_KEY_set_private_key(key.get(), private_bignum)) {
+      return FAILED;
+    }
+
+    // EC_KEY that has a private key but do not has public key assumed invalid
+    // thus we are generating its public key part
+    EC_POINT *public_key_point{EC_POINT_new(curve)};
+    if (nullptr == public_key_point) {
+      return FAILED;
+    }
+    auto free_public_key_point =
+        gsl::finally([public_key_point] { EC_POINT_free(public_key_point); });
+
+    // derive the public key
+    if (1
+        != EC_POINT_mul(curve, public_key_point, private_bignum, nullptr,
+                        nullptr, nullptr)) {
+      return FAILED;
+    }
+
+    // check the public key
+    if (1 != EC_POINT_is_on_curve(curve, public_key_point, nullptr)) {
+      return KeyValidatorError::INVALID_PUBLIC_KEY;
+    }
+
+    // set public key to the resulting key structure
+    if (1 != EC_KEY_set_public_key(key.get(), public_key_point)) {
+      return FAILED;
+    }
+
+    // final check
+    if (1 != EC_KEY_check_key(key.get())) {
+      return KeyValidatorError::INVALID_PRIVATE_KEY;
+    }
+
+    return key;
+  }
+
+  template <typename ReaderFunc>
+  outcome::result<std::shared_ptr<EVP_PKEY>> NewEvpPkeyFromBytes(
+      int type, gsl::span<const uint8_t> key_bytes, ReaderFunc *reader) {
```

Comment:
the `reader` name is not obvious, could you pls explain why you use it?

Reply:
1 It is better than typename T
2 It is a pointer to a function that does read of bytes and interpret them as a public or private key

Now it has docstring that contains examples.